### ISSUE TITLE
[BUG] apply_boolean_mask misbehaves when inputs are not nullable

### ISSUE THREAD

**Issue by williamBlazing** [rapidsai/cudf#3714]
**Describe the bug**
`apply_boolean_mask`  is giving the wrong results for certain small width data types and the output bitmask is filled for strings but not other data types

**Steps/Code to reproduce bug**
`template <typename T>
struct ApplyFilter : public cudf::test::BaseFixture {};

TYPED_TEST_CASE(ApplyFilter, cudf::test::NumericTypes);
TYPED_TEST(ApplyFilter, withNull)
{
    using T = TypeParam;
    cudf::test::fixed_width_column_wrapper<T> col1{{5, 4, 3, 5, 8, 5, 6}, {1, 1, 0, 1, 1, 1, 1}};
    cudf::test::strings_column_wrapper col2({"d", "e", "a", "d", "k", "d", "l"}, {1, 0, 1, 1, 1, 1, 1});
    cudf::test::fixed_width_column_wrapper<T> col3{{10, 40, 70, 5, 2, 10, 11}, {0, 1, 1, 1, 1, 1, 0}};
    cudf::table_view cudf_table_in_view {{col1, col2, col3}};

    cudf::test::fixed_width_column_wrapper<cudf::experimental::bool8> bool_filter{{1, 1, 1, 0, 1, 0, 0}, {1, 1, 0, 1, 1, 1, 0}};
    cudf::column_view bool_filter_col(bool_filter);

    std::unique_ptr<cudf::experimental::table> filteredTable = cudf::experimental::apply_boolean_mask(cudf_table_in_view,bool_filter_col);
    cudf::table_view tableView = filteredTable->view();
    
    cudf::test::fixed_width_column_wrapper<T> expect_col1{{5, 4, 8}, {1, 1,1}};
    cudf::test::strings_column_wrapper expect_col2({"d", "e", "k"}, {1, 0, 1});
    cudf::test::fixed_width_column_wrapper<T> expect_col3{{10, 40, 2}, {0, 1, 1}};
    cudf::table_view expect_cudf_table_view {{expect_col1, expect_col2, expect_col3}};

    std::string col0_string = cudf::test::to_string(tableView.column(0), "|");
    if (tableView.column(0).nullable())
        col0_string = col0_string + " is nullable";
    std::cout<<"col0_string: "<<col0_string<<std::endl;    
    std::string col1_string = cudf::test::to_string(tableView.column(1), "|");
    if (tableView.column(1).nullable())
        col1_string = col1_string + " is nullable";
    std::cout<<"col1_string: "<<col1_string<<std::endl;
    std::string col2_string = cudf::test::to_string(tableView.column(2), "|");
    if (tableView.column(2).nullable())
        col2_string = col2_string + " is nullable";
    std::cout<<"col2_string: "<<col2_string<<std::endl;

    cudf::test::expect_tables_equal(expect_cudf_table_view, tableView);
}

TYPED_TEST(ApplyFilter, withoutNull)
{
    using T = TypeParam;
    cudf::test::fixed_width_column_wrapper<T> col1{{5, 4, 3, 5, 8, 5, 6}};
    cudf::test::strings_column_wrapper col2({"d", "e", "a", "d", "k", "d", "l"});
    cudf::test::fixed_width_column_wrapper<T> col3{{10, 40, 70, 5, 2, 10, 11}};
    cudf::table_view cudf_table_in_view {{col1, col2, col3}};

    cudf::test::fixed_width_column_wrapper<cudf::experimental::bool8> bool_filter{{1, 1, 0, 0, 1, 0, 0}};
    cudf::column_view bool_filter_col(bool_filter);

    std::unique_ptr<cudf::experimental::table> filteredTable = cudf::experimental::apply_boolean_mask(cudf_table_in_view,bool_filter_col);
    cudf::table_view tableView = filteredTable->view();
    
    cudf::test::fixed_width_column_wrapper<T> expect_col1{{5, 4, 8}};
    cudf::test::strings_column_wrapper expect_col2({"d", "e", "k"});
    cudf::test::fixed_width_column_wrapper<T> expect_col3{{10, 40, 2}};
    cudf::table_view expect_cudf_table_view {{expect_col1, expect_col2, expect_col3}};

    std::string col0_string = cudf::test::to_string(tableView.column(0), "|");
    if (tableView.column(0).nullable())
        col0_string = col0_string + " is nullable";
    std::cout<<"col0_string: "<<col0_string<<std::endl;    
    std::string col1_string = cudf::test::to_string(tableView.column(1), "|");
    if (tableView.column(1).nullable())
        col1_string = col1_string + " is nullable";
    std::cout<<"col1_string: "<<col1_string<<std::endl;
    std::string col2_string = cudf::test::to_string(tableView.column(2), "|");
    if (tableView.column(2).nullable())
        col2_string = col2_string + " is nullable";
    std::cout<<"col2_string: "<<col2_string<<std::endl;

    cudf::test::expect_tables_equal(expect_cudf_table_view, tableView);
}`

**Expected behavior**
The unit test provided above should pass

**Environment overview (please complete the following information)**
Using cudf nightly from 01/06/20 installed using conda

**Environment details**
Please run and paste the output of the `cudf/print_env.sh` script here, to gather any other relevant environment details

**Additional context**
Unit test originally was run in a combination of source code and conda installed library. Bug was discovered in BlazingSQL unit test and was then distilled to only the cudf code.

**Comment by jrhemstad** [comment]
@rgsl888prabhu can you triage this?

**Comment by rgsl888prabhu** [comment]
By default gather was creating null_mask, which caused this issue. I will be pushing a PR soon.

**Comment by jlowe** [comment]
We ran into this as well on the Spark side.  A small test case of filtering just one integer column and involving no strings fails:
```c++
TEST_F(ApplyBooleanMask, BadFilterTest) {
  cudf::test::fixed_width_column_wrapper<int> col({9668, 9590, 9526, 9205, 9434, 9347, 9160, 9569, 9143, 9807, 9606, 9446, 9279, 9822, 9691});
  cudf::test::fixed_width_column_wrapper<cudf::experimental::bool8> mask({false, false, true, false, false, true, false, true, false, true, false, false, true, false, true});
  cudf::table_view input({col});
  cudf::test::fixed_width_column_wrapper<int> col_expected({9526,9347,9569,9807,9279,9691});
  cudf::table_view expected({col_expected});
  auto got = cudf::experimental::apply_boolean_mask(input, mask);
  cudf::test::print(got->get_column(0), std::cout, ",");
  cudf::test::expect_tables_equal(expected, got->view());
}
```

This continues to fail after applying the patch from #3278.

**Comment by rgsl888prabhu** [comment]
> We ran into this as well on the Spark side. A small test case of filtering just one integer column and involving no strings fails:
> 
> ```c++
> TEST_F(ApplyBooleanMask, BadFilterTest) {
>   cudf::test::fixed_width_column_wrapper<int> col({9668, 9590, 9526, 9205, 9434, 9347, 9160, 9569, 9143, 9807, 9606, 9446, 9279, 9822, 9691});
>   cudf::test::fixed_width_column_wrapper<cudf::experimental::bool8> mask({false, false, true, false, false, true, false, true, false, true, false, false, true, false, true});
>   cudf::table_view input({col});
>   cudf::test::fixed_width_column_wrapper<int> col_expected({9526,9347,9569,9807,9279,9691});
>   cudf::table_view expected({col_expected});
>   auto got = cudf::experimental::apply_boolean_mask(input, mask);
>   cudf::test::print(got->get_column(0), std::cout, ",");
>   cudf::test::expect_tables_equal(expected, got->view());
> }
> ```
> 
> This continues to fail after applying the patch from #3728.

This has been resolved in PR #3681, check this comment. This check should get that test case pass. I will move that change to this PR to get it moving. https://github.com/rapidsai/cudf/pull/3681#discussion_r365031299

**Comment by jlowe** [comment]
Thanks, @rgsl888prabhu!  I verified pulling the latest PR changes fixes the queries that were previously failing.
